### ISSUE TITLE
Derivatives of input no fatal error

### ISSUE THREAD

**Issue by AnHeuermann** [OpenModelica/OpenModelica#9802]
### Description

When translating a model where an input needs to be differentiated omc reports an error but won't stop the translation. Instead it fails to compile the garbage C code later on.

### Steps to Reproduce

```Modelica
model illegalDerInput
  input Real u;
  output Real y;
equation
  y = der(u);
end illegalDerInput;
```

```
loadFile("illegalDerInput.mo"); getErrorString();
buildModel(illegalDerInput); getErrorString();
```

Error message:
```bash
Error: The model requires derivatives of some inputs as listed below:
der(u)
[...]
illegalDerInput_06inz.c:20:56: error: use of undeclared identifier '_omcQ_24DER'
  (data->localData[0]->realVars[1] /* y variable */) = _omcQ_24DER._u;
                                                       ^
1 error generated.
```

### Expected Behavior

Brake at the first reported error and don't continue.

### Version and OS

 - OpenModelica Version: v1.21.0-dev-77-gf539850045-cmake
 - OS: Ubuntu 20.04.5 LTS

**Comment by casella** [comment]
Isn't this maybe too restrictive? If the input is provided by CSV files, it is interpreted as a piecewise linear function, so we could compute the derivative by looking at the slope of the piecewise-linear function.

We should just generate a runtime error if the input has a step change (i.e., two different values for the same time value).

For FMI, I'm not sure, maybe it would be wiser to avoid this, because FMUs can go back and forth in time, and figuring out the derivative consistently may be tricky.

**Comment by AnHeuermann** [comment]
> Isn't this maybe too restrictive? 

I was a bit imprecise. I mean top-level inputs.
But top-level inputs from a CSV file seem to be an issue.
Try simulating `illegalDerInput` with `-simflags=input.csv` for `time^2` inputs:

```CSV
"time","illegalDerInput.u"
0,0
0.002,4e-06
0.004,1.6e-05
0.006,3.6e-05
0.008,6.4e-05
0.01,0.0001
0.012,0.000144
[...]
```

Fixing this is outside the scope of this issue, but probably a good idea.

But for now getting translation error

```
Error: The model requires derivatives of some inputs as listed below:
der(u)
Error: post-optimization module inputDerivativesUsed (simulation) failed.
```

is better than the C compilation error:
```
[...]
illegalDerInput.c:106:56: error: use of undeclared identifier '_omcQ_24DER'
  (data->localData[0]->realVars[1] /* y variable */) = _omcQ_24DER._u;
                                                       ^
illegalDerInput_06inz.c:20:56: error: use of undeclared identifier '_omcQ_24DER'
  (data->localData[0]->realVars[1] /* y variable */) = _omcQ_24DER._u;
                                                       ^
1 error generated.
1 error generated.
mingw32-make: *** [<builtin>: illegalDerInput_06inz.o] Error 1
mingw32-make: *** Waiting for unfinished jobs....
mingw32-make: *** [<builtin>: illegalDerInput.o] Error 1
mingw32-make: Leaving directory 'C:/Users/AHEUER~1/AppData/Local/Temp/OPENMO~1/OMEdit/ILLEGA~1'
Compilation process failed. Exited with code 0x00000002.
```

**Comment by AnHeuermann** [comment]
> For FMI, I'm not sure, maybe it would be wiser to avoid this, because FMUs can go back and forth in time, and figuring out the derivative consistently may be tricky.

This is a completely different question. Not sure what the specification has to say about this. But e.g. for Co-Simulation FMUs with communication points I don't see a way to get the derivative information from a different FMU/CSV file, at least not for the exported FMU. The FMU importing tool may be able to provide this information by numerically differentiating. But if the communication points are to far apart this could be difficult.

**Comment by casella** [comment]
As I understand so far, in principle there is nothing wrong in considering the derivative of top-level inputs, since they are not constant in general. So, I would suggest that the backend takes them into account if needed, rather than arbitrarily (and incorrectly) assuming those derivatives are zero.

How these derivative end up being handled in the generated code, it depends on the context:
- for simulation without a csv input file, those inputs are assumed to be equal to their start value, so their derivatives should be computed to zero at initialization
- for simulation with a csv input file, we should fix the code to get a correct derivative. I would use the right limit, so the derivative is well-defined at StartTime. If at StopTime there are no further data available, we can use the left limit
- for FMI-CS, see discussion in #9782, the input derivatives could be provided via `fmi2SetRealInputDerivatives`.

Of course we shouldn't be generating garbage code as we do today 😄

**Comment by AnHeuermann** [comment]
I agree, but all of these are different issues, so I opened one for the CSV case.

**Comment by casella** [comment]
Test case reported by @MartinOtter in [#14698](https://github.com/OpenModelica/OpenModelica/issues/14698#issuecomment-3580470339):

```modelica
model TestDifferentiation5
  Modelica.Blocks.Interfaces.RealInput u
    annotation (Placement(transformation(extent={{-140,-20},{-100,20}})));
  Modelica.Blocks.Math.InverseBlockConstraints inverseBlockConstraints
    annotation (Placement(transformation(extent={{-74,-12},{-34,12}})));
  Modelica.Blocks.Continuous.SecondOrder secondOrder(w=1, D=1)
    annotation (Placement(transformation(extent={{-44,-10},{-64,10}})));
equation 
  connect(u, inverseBlockConstraints.u1)
    annotation (Line(points={{-120,0},{-76,0}}, color={0,0,127}));
  connect(secondOrder.y, inverseBlockConstraints.u2)
    annotation (Line(points={{-65,0},{-70,0}}, color={0,0,127}));
  connect(inverseBlockConstraints.y2, secondOrder.u)
    annotation (Line(points={{-37,0},{-42,0}}, color={0,0,127}));
  annotation (
    Icon(coordinateSystem(preserveAspectRatio=false)),
    Diagram(coordinateSystem(preserveAspectRatio=false)),
    uses(Modelica(version="4.0.0")));
end TestDifferentiation5;
```
This model requires to differentiate the input twice, so it should fail at compile time. In fact, it doesn't: the compilation process assumes the input is identically zero and produces simulation code accordingly.

**Comment by casella** [comment]
Here is an MWE of the above case
```modelica
model MWE
  input Real u_top;
  Real u;
  Real y;
  parameter Real tau = 1;
equation
  tau*der(y) + y = u "First-order low-pass filter";
  y = u_top "The input prescribes the output of the filter, inverse problem";
end MWE;
```
@AnHeuermann I guess the problem is that the fix in #9804 only works in trivial cases where the input is _explicitly_ differentiated, e.g.

```modelica
model MWE
  input Real u;
  Real y;
equation
  y = der(u);
end MWE;
```

**Comment by casella** [comment]
Keeping @MartinOtter in the loop.
